### PR TITLE
update JS SDK changelog tool

### PR DESCRIPTION
**Pull request by kazrael2119**
https://github.com/Azure/azure-sdk-tools/issues/7883#issuecomment-1998841283
fixes https://github.com/Azure/azure-sdk-for-js/issues/26543

### CODE REVIEW

File: tools/js-sdk-release-tools/src/hlc/utils/modifyChangelogFileAndBumpVersion.ts [hunk 1]
```diff
@@ -17,7 +17,7 @@ export function makeChangesForFirstRelease(packageFolderPath: string, isStableRe
     
 ## ${newVersion} (${date})
 
-The package of ${packageJsonData.name} is using our next generation design principles. To learn more, please refer to our documentation [Quick Start](https://aka.ms/js-track2-quickstart).
+The package of ${packageJsonData.name} is using our next generation design principles. To learn more, please refer to our documentation [Quick Start](https://aka.ms/azsdk/js/mgmt/quickstart ).
```
Comment (MaryGao) [inline tools/js-sdk-release-tools/src/hlc/utils/modifyChangelogFileAndBumpVersion.ts:20]: Steps for changelog tools:
- Modify the change under folder `js-sdk-release-tools`
- Build in local version and verify in sdk repo
- Bump the version in sdk-tools repo in packge.json
- Upgrade the lock file also

**Comment by kazrael2119** [comment]
need test :
pure stable -> stable
pure preview -> stable
mixed stable -> preview
mixed preview -> preview

File: tools/js-sdk-release-tools/src/hlc/utils/modifyChangelogFileAndBumpVersion.ts [hunk 2]
```diff
@@ -62,11 +62,8 @@ function changeClientFile(packageFolderPath: string, packageVersion: string) {
     })
 }
 
-export function makeChangesForReleasingTrack2(packageFolderPath: string, packageVersion: string, changeLog: Changelog) {
-    let originalChangeLogContent = fs.readFileSync(path.join(packageFolderPath, 'changelog-temp', 'package', 'CHANGELOG.md'), {encoding: 'utf-8'});
-    if(originalChangeLogContent.includes("https://aka.ms/js-track2-quickstart")){
-        originalChangeLogContent=originalChangeLogContent.replace("https://aka.ms/js-track2-quickstart","https://aka.ms/azsdk/js/mgmt/quickstart");
-    }
+export function makeChangesForReleasingTrack2(packageFolderPath: string, packageVersion: string, changeLog: Changelog, changeLogContent: string) {
```
Comment (kazrael2119) [inline tools/js-sdk-release-tools/src/hlc/utils/modifyChangelogFileAndBumpVersion.ts:65]:
```suggestion
export function makeChangesForReleasingTrack2(packageFolderPath: string, packageVersion: string, changeLog: Changelog, originalChangeLogContent: string) {
```

**Comment by kazrael2119** [comment]
support compared information if the new pacakge is a beta version
image

File: tools/js-sdk-release-tools/src/utils/version.ts [hunk 3]
```diff
@@ -2,12 +2,23 @@ import {NPMViewResult, StringMap, tr} from "@ts-common/azure-js-dev-tools";
 import {logger} from "./logger";
 const semverInc = require('semver/functions/inc')
 
+export function getVersion(npmViewResult: NPMViewResult,Tag: string) {
+    const distTags: StringMap<string> | undefined = npmViewResult['dist-tags'];
+    const Version = distTags && distTags[Tag];
```
Comment (MaryGao) [inline tools/js-sdk-release-tools/src/utils/version.ts:7]: what is your naming style? usually we use Camel case naming for variable.

File: tools/js-sdk-release-tools/src/utils/version.ts [hunk 4]
```diff
@@ -2,12 +2,23 @@ import {NPMViewResult, StringMap, tr} from "@ts-common/azure-js-dev-tools";
...
+    return Version;
+}
+
+export function getversionDate(npmViewResult: NPMViewResult, Version : string){
+    const time: StringMap<string> | undefined = npmViewResult['time'];
+    const VersionDate = time && time[Version];
```
Comment (MaryGao) [inline tools/js-sdk-release-tools/src/utils/version.ts:13]: I don't think we need extra var here `VersionDate`

File: tools/js-sdk-release-tools/src/utils/version.ts [hunk 5]
```diff
@@ -2,12 +2,23 @@ import {NPMViewResult, StringMap, tr} from "@ts-common/azure-js-dev-tools";
...
+    return VersionDate;
```
Comment (MaryGao) [inline tools/js-sdk-release-tools/src/utils/version.ts:14]:
```suggestion
    return time && time[Version];
```

File: tools/js-sdk-release-tools/src/hlc/utils/automaticGenerateChangeLogAndBumpVersion.ts [hunk 6]
```diff
@@ -57,6 +60,27 @@ export async function generateChangelogAndBumpVersion(packageFolderPath: string)
...
+                if(nextVersion){
+                    await shell.cd(path.join(packageFolderPath, 'changelog-temp'));
+                    await shell.mkdir(path.join(packageFolderPath, 'changelog-temp', 'changelog-next'));
+                    await shell.cd(path.join(packageFolderPath,'changelog-temp', 'changelog-next'));
+                    await shell.exec(`npm pack ${packageName}@${nextVersion}`);
+                    await shell.exec('tar -xzf *.tgz');
+                    await shell.cd(packageFolderPath);
+                    logger.log("Create next folder successfully")
+    
+                    const LatestDate = getversionDate(npmViewResult, stableVersion);
```
Comment (MaryGao) [inline tools/js-sdk-release-tools/src/hlc/utils/automaticGenerateChangeLogAndBumpVersion.ts:73]: naming style

File: tools/js-sdk-release-tools/src/hlc/utils/automaticGenerateChangeLogAndBumpVersion.ts [hunk 7]
```diff
@@ -57,6 +60,27 @@ export async function generateChangelogAndBumpVersion(packageFolderPath: string)
                 let apiMdFileNPM: string = path.join(reviewFolder, fs.readdirSync(reviewFolder)[0]);
                 let apiMdFileLocal: string = path.join(packageFolderPath, 'review', fs.readdirSync(path.join(packageFolderPath, 'review'))[0]);
                 const changelog: Changelog = await extractExportAndGenerateChangelog(apiMdFileNPM, apiMdFileLocal);
+                let originalChangeLogContent = fs.readFileSync(path.join(packageFolderPath, 'changelog-temp', 'package', 'CHANGELOG.md'), {encoding: 'utf-8'});
+                if(nextVersion){
+                    await shell.cd(path.join(packageFolderPath, 'changelog-temp'));
+                    await shell.mkdir(path.join(packageFolderPath, 'changelog-temp', 'changelog-next'));
```
Comment (MaryGao) [inline tools/js-sdk-release-tools/src/hlc/utils/automaticGenerateChangeLogAndBumpVersion.ts:66]: changelog-next -> next

File: tools/js-sdk-release-tools/src/hlc/utils/automaticGenerateChangeLogAndBumpVersion.ts [hunk 8]
```diff
@@ -57,6 +60,27 @@ export async function generateChangelogAndBumpVersion(packageFolderPath: string)
...
+                    await shell.cd(path.join(packageFolderPath,'changelog-temp', 'changelog-next'));
+                    await shell.exec(`npm pack ${packageName}@${nextVersion}`);
+                    await shell.exec('tar -xzf *.tgz');
+                    await shell.cd(packageFolderPath);
+                    logger.log("Create next folder successfully")
+    
+                    const LatestDate = getversionDate(npmViewResult, stableVersion);
+                    const NextDate = getversionDate(npmViewResult,nextVersion);
+                    if (LatestDate && NextDate){
```
Comment (MaryGao) [inline tools/js-sdk-release-tools/src/hlc/utils/automaticGenerateChangeLogAndBumpVersion.ts:75]:
```suggestion
                    if (LatestDate && NextDate && LatestDate <= NextDate)) {
```

File: tools/js-sdk-release-tools/src/hlc/utils/automaticGenerateChangeLogAndBumpVersion.ts [hunk 9]
```diff
@@ -57,6 +60,27 @@ export async function generateChangelogAndBumpVersion(packageFolderPath: string)
...
+                        if (LatestDate <= NextDate){
+                            originalChangeLogContent = fs.readFileSync(path.join(packageFolderPath,'changelog-temp', 'changelog-next', 'package', 'CHANGELOG.md'), {encoding: 'utf-8'});
```
Comment (MaryGao) [inline tools/js-sdk-release-tools/src/hlc/utils/automaticGenerateChangeLogAndBumpVersion.ts:77]: Need to keep previous preview changelog

**Comment by kazrael2119** [comment]
no issues if it is a firstGA package:
image

**Review by MaryGao — APPROVED** [review]
LGTM

**Pull request merged**